### PR TITLE
Add macOS and FreeBSD support and fix DJGPP build

### PR DESCRIPTION
**Pull request by techguy16**
Sorry about almost 50 commits, it was me experimenting (sorry).

### CODE REVIEW

**Comment by thepwrtank18** [comment]
Anyone have a Mac to test this?

**Comment by techguy16** [comment]
I added a test in one of the workflow sections.

**Comment by techguy16** [comment]
Proof the Mac build works.

File: CMakeLists.txt
Comment (Neo-Desktop) [inline CMakeLists.txt:1]: is there any particular reason why you lower cased the function calls?

File: README.md
Comment (Neo-Desktop) [inline README.md:1]: thank you for updating the readme

File: README.md [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+
+**Build status**
 [![C/C++ CI (Linux)](https://github.com/UMSKT/UMSKT/actions/workflows/linux.yml/badge.svg)](../../actions/workflows/linux.yml)
 [![C/C++ CI (Windows)](https://github.com/UMSKT/UMSKT/actions/workflows/windows.yml/badge.svg)](../../actions/workflows/windows.yml)
+[![C/C++ CI (macOS)](https://github.com/UMSKT/UMSKT/actions/workflows/macos.yml/badge.svg)](../../actions/workflows/macos.yml)
+[![C/C++ CI (DOS DJGPP)](https://github.com/UMSKT/UMSKT/actions/workflows/dos-djgpp.yml/badge.svg)](../../actions/workflows/freebsd.yml)
+[![C/C++ CI (FreeBSD)](https://github.com/UMSKT/UMSKT/actions/workflows/freebsd.yml/badge.svg)](../../actions/workflows/dos-djgpp.yml)
 
 
-### **Plan of Action / ToDo List**
```
Comment (Neo-Desktop) [inline README.md:9]: any reason why you removed this? those threads are more or less up to date still

File: README.md [hunk 2]
```diff
@@ -31,25 +32,25 @@ In light of the recent exponential interest in this project I've decided to put
...
-    * Download the latest experimental version using the Actions tab ([Windows](../../actions/workflows/windows.yml?query=branch%3Amaster), [Linux](../../actions/workflows/linux.yml?query=branch%3Amaster)).
+    * Download the latest experimental version using the Actions tab ([Windows](../../actions/workflows/windows.yml?query=branch%3Amaster), [Linux](../../actions/workflows/linux.yml?query=branch%3Amaster)), [macOS](../../actions/workflows/macos.yml?query=branch%3Amaster), [FreeBSD](../../actions/workflows/freebsd.yml?query=branch%3Amaster), [DOS DJGPP](../../actions/workflows/.yml?qudos-djgppery=branch%3Amaster))
 
 
 * *(GitHub account \*not\* required)*
     * Download the latest release for your operating system and architecture from [the releases page](../../releases)
 
 
-* **Note:** Before continuing, please ensure you have both the `xpkey` program and the `keys.json` datum extracted and in the same directory
+* **Note:** Before continuing, please ensure you have the `umskt` executablw extracted.
```
Comment (Neo-Desktop) [inline README.md:45]: typo: should be `executable`

File: README.md [hunk 3]
```diff
@@ -68,7 +69,7 @@ The list of people who have helped to bring the XP generation to where it is now
 * WitherOrNot
 * TheTank20
 * InvoxiPlayGames
-* brakmic
+* bra
```
Comment (Neo-Desktop) [inline README.md:72]: why was this changed?

File: src/cli.cpp [hunk 4]
```diff
@@ -241,7 +241,7 @@ void CLI::printID(DWORD *pid) {
     int i, digit = 0;
 
     // Convert PID to ascii-number (=raw)
-    sprintf(raw, "%09u", pid[0]);
+    snprintf(raw, sizeof(raw), "%09u", pid[0]);
```
Comment (Neo-Desktop) [inline src/cli.cpp:244]: ah yes, we should've be using snprintf here

perhaps when we switch from fmt to pocoproject this will be a good candidate to abstract

File: .github/workflows/dos-djgpp.yml [hunk 5]
```diff
@@ -39,7 +43,7 @@ jobs:
       - name: Setup build environment
         run: |
           sudo apt -y update
-          sudo apt -y install build-essential cmake wget 7zip git flex nasm libslang2-dev pkg-config libslang2-modules gcc-multilib
+          sudo apt -y install build-essential cmake wget 7zip git flex nasm libslang2-dev pkg-config libslang2-modules gcc-multilib g++
```
Comment (Neo-Desktop) [inline .github/workflows/dos-djgpp.yml:46]: isn't `g++` part of `build-essential`?

**Review by Neo-Desktop — CHANGES_REQUESTED** [review]
needs some changes but looks good

File: .github/workflows/dos-djgpp.yml (hunk 5 above)
Comment (techguy16) [reply .github/workflows/dos-djgpp.yml:46]: No.

File: README.md (hunk 3 above)
Comment (techguy16) [reply README.md:72]: Oops, coding on my phone 🤣

Comment (thepwrtank18) [reply README.md:72]: ~~based~~

File: README.md (hunk 1 above)
Comment (techguy16) [reply README.md:9]: Fixed it.

File: .github/workflows/dos-djgpp.yml (hunk 5 above)
Comment (Neo-Desktop) [reply .github/workflows/dos-djgpp.yml:46]: https://packages.ubuntu.com/mantic/build-essential

`g++` is definitely a dependency of `build-essential`

image

File: CMakeLists.txt
Comment (techguy16) [reply CMakeLists.txt:1]: Cleaner?

File: .github/workflows/dos-djgpp.yml (hunk 5 above)
Comment (techguy16) [reply .github/workflows/dos-djgpp.yml:46]: But it doesn't install in it.

**Comment by thepwrtank18** [comment]
Me waiting for a new commit for me to click the "Approve and run" button
monkey-toy-story3

**Comment by techguy16** [comment]
🤣

**Comment by Neo-Desktop** [comment]
yeah we don't need it to rebuild the project when we update the readme - I'm gonna see if I can fix that right now

File: .github/workflows/freebsd.yml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+on:
+  push:
+    branches: [ "master" ]
+  pull_request:
+    branches: [ "master" ]
+
+jobs:
+  build:
+    runs-on: macos-12
```
Comment (Neo-Desktop) [inline .github/workflows/freebsd.yml:31]: why is this running on macos?

File: .github/workflows/freebsd.yml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    name: build-x86_64
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Build & Test in FreeBSD
+      id: test
+      uses: vmactions/freebsd-vm@v0
+      with:
+        envs: 'MYTOKEN MYTOKEN2'
```
Comment (Neo-Desktop) [inline .github/workflows/freebsd.yml:40]: is this needed?

Comment (Neo-Desktop) [reply .github/workflows/freebsd.yml:40]: this is not needed. I'll remove it

File: .github/workflows/freebsd.yml (hunk 6 above)
Comment (Neo-Desktop) [reply .github/workflows/freebsd.yml:31]: the macos runner appears to have virtualbox installed by default, which is the easiest way to compile for freebsd it appears

Comment (Neo-Desktop) [reply .github/workflows/freebsd.yml:31]: ah wait, I've found a better solution
https://github.com/marketplace/actions/cross-platform-action

File: .github/workflows/freebsd.yml (hunk 7 above)
Comment (techguy16) [reply .github/workflows/freebsd.yml:40]: Oops...

File: .github/workflows/freebsd.yml (hunk 6 above)
Comment (techguy16) [reply .github/workflows/freebsd.yml:31]: No the macOS + VB is better.

File: .github/workflows/freebsd.yml (hunk 7 above)
Comment (techguy16) [reply .github/workflows/freebsd.yml:40]: You can fix them if you want.

**Review by techguy16 — COMMENTED** [review]
Cool!
